### DISCUSSION TITLE
Deploy Milvus Cluster without Helm/K8s, Volume and Network issue?

### DISCUSSION THREAD

**Discussion by kakkoiiAllen** [milvus-io/milvus#18257] [Q&A and General discussion]
Hi:
I'm an intern trying to deploy the cluster version using Nomad instead of Helm/K8s, but I have some questions/problems regarding to the deployment and I'd really appreciate your help. (plz excuse me for stupid question as I know little about distributed system)

1. MinIO and ETCD are for storing data/log, and we need to have multiple(duplicate) nodes of them mounting to the same volume to ensure data persistence? For instance we have minio-01, minio-02, minio-03, they are basically duplication nodes, but mounting to the same /minio/data folder of the host so that data persists even one node is dead?
2. I have a hard time understanding the network/communication between nodes. From docker-compose file, each worker node(index coord, index node, data coord, data node, .etc) only have env variable for ETCD, MinIO, and Pulsar's server address, but the networking shows that there're other network connection between these nodes. Query node has network connection to root coord and query coord, proxy node has connection to other coord nodes. I couldn't find the config of these network settings. Where to set these local address and the foreign address? Do I need to modify the codebase?
<img width="566" alt="Screen Shot 2022-07-12 at 9 50 21 PM" src="https://user-images.githubusercontent.com/94626600/178653222-652bd88d-669b-4ab1-94c0-36ae5be44326.png">
Since I want to deploy the system through Nomad, I would need to know how to connect all these nodes together through ip address. I can get the ip address of MinIO, ETCD, Pulsar server and assign the corresponding envvar, but I don't know how to do the rest of the network connection. May anyone give any advice? Thank you so much!

Update:
I just found that ETCD stores the necessary ip:port info, and the network connection may setup automatically as long as the ETCD envvar is correct? I tried to build node separately and set up envvars, but I got some problem connecting them together. I set up ETCD, MinIO, Pulsar-standalone, and now I'd like to start milvus-rootcoord, I already checked the env var and they are correct, but I got this error below:
<img width="1342" alt="Screen Shot 2022-07-13 at 6 33 30 PM" src="https://user-images.githubusercontent.com/94626600/178883202-de7c2b2c-21b0-42a1-9972-d8e1949f6076.png">
The rootcoord service won't start. I tried the same thing locally through docker compose and start service one by one, and it run without problem. Any idea why it is the case? Am I on the right track?

**Comment by yhmo** [comment]
Actually, we have no experience on the Nomad tool. Helm/K8s is the recommend way to deploy milvus cluster. Sometimes we use docker compose to deploy milvus cluster on a single machine, mainly for demo and testing. You must do special configuration for docker-compose to deploy milvus on multiple machines, and once a node is down, you always need to manually restart it, indeed it is not a good idea to do that.

@LoveEachDay  Do you know the question of etcd?

**Reply by kakkoiiAllen** [reply]
I have another question. Is it possible to change the ip:port address of the services. For example, I have a rootcoord and it is connected to ETCD, I don't want ETCD to register the "localhost" ip address (192.168 / 172.xxx), but the public ip address that is accessible to other nodes, how can I do it? Can I do this through envvars or milvus/config? Can I setup an envvar and change milvus/config rootcoord address to something like ${rootcoord_address}?

**Reply by LoveEachDay** [reply]
You can manually modify the value in the etcd. When the milvus component starts up, if it starts normally, it will register itself into the etcd. You can following the directions here to find it out:
1. Log into the etcd container or pod
2. Execute the following the command: `etcdctl get --prefix by-dev/meta/session`. You'll get something like this:
```
by-dev/meta/session/rootcoord
{"ServerID":1,"ServerName":"rootcoord","Address":"10.104.5.116:53100","Exclusive":true,"TriggerKill":true}
```
4. Take `rootcoord` as an example, change the `by-dev/meta/session/rootcoord` like this:
```
etcdctl put by-dev/meta/session/rootcoord '{"ServerID":1,"ServerName":"rootcoord","Address":"<new-ip>:53100","Exclusive":true,"TriggerKill":true}'
```

Note: You can do this manually, but we don't suggest user to do this. The metadata is very delicate, if you mess it up, your milvus cluster will behave differently.

**Reply by kakkoiiAllen** [reply]
Thanks! I'm also trying to build image from source to explore more about Milvus. I tried builder.sh then build_image.sh, but I got /lib not found, am I missing anything? 

For the ETCD part, it is assumed that all coords and nodes to be on the same network? Is it not suggested to put these nodes onto different network? I ask the question because the Nomad framework I use cannot guarantee services to be on the same network, and thus cannot talk to each other using private ip address.
